### PR TITLE
fix(cli): fix overwriting dataset tags

### LINKED ISSUE #3463: cli: renku dataset tag dataset-name dataset- [...]
## Describe the bug
I was working with tags for datasets, and wanted to overwrite an already existing tag (the goal was to change the tag's description by using the following command:
`renku dataset tag dataset-name dataset-tag -f -d 'This is a short description'`

## Details
The automatically created bug report is as follows:

**Renku version:** 2.4.1rc1
**OS:** Darwin (Darwin Kernel Version 22.1.0: Sun Oct  9 20:19:12 PDT 2022; root:xnu-8792.41.9~3/RELEASE_ARM64_T6020)
**Python:** 3.9.6

### Traceback

```
Traceback (most recent call last):
  File "[...]/renku/command/command_builder/database.py", line 110, in _post_hook
    project_context.pop_context()
  File "[...]/renku/domain_model/project_context.py", line 247, in pop_context
    self._top.database.commit()
  File "[...]/renku/infrastructure/database.py", line 459, in commit
    self._store_object(object)
  File "[...]/renku/infrastructure/database.py", line 466, in _store_object
    self._cache[object._p_oid] = object
  File "[...]/renku/infrastructure/database.py", line 532, in __setitem__
    raise ValueError(f"The same oid exists: {existing_data} != {object}")
ValueError: The same oid exists: <renku.domain_model.dataset.DatasetTag object at 0x1073c92e0 oid '9ad0666be7e40f11eceec79a004d2364cea7ab9a735bca66e72db03d6f893437' in <renku.infrastructure.database.Database object at 0x1073a8160>> != <renku.domain_model.dataset.DatasetTag object at 0x1073c93c0 oid '9ad0666be7e40f11eceec79a004d2364cea7ab9a735bca66e72db03d6f893437' in <renku.infrastructure.database.Database object at 0x1073a8160>>
```

## Additional context
Add any other context about the problem.

### PR DESCRIPTION
**Pull request by Panaetius**
closes #3463

### CODE REVIEW

File: renku/infrastructure/gateway/dataset_gateway.py [hunk 1]
```diff
@@ -69,6 +69,7 @@ def remove_tag(self, dataset: Dataset, tag: DatasetTag):
         for t in tags:
             if t.name == tag.name:
                 tags.remove(t)
+                project_context.database._cache.pop(t._p_oid)
```
Comment (mohammad-alisafaee) [inline renku/infrastructure/gateway/dataset_gateway.py:72]: I believe it's cleaner if we add a method named `invalidate_cache` or something similar to the database and call it here. It also make it easier to look for all such uses.

File: tests/cli/test_datasets.py [hunk 2]
```diff
@@ -1547,6 +1547,38 @@ def test_dataset_tag(tmpdir, runner, project, subdirectory):
...
+    assert 0 == result.exit_code, format_result_exception(result)
+    assert "OK" in result.output
+
+    # create some data
+    new_file = tmpdir.join("file")
+    new_file.write("test")
+
+    # add data to dataset
+    result = runner.invoke(cli, ["dataset", "add", "--copy", "my-dataset", str(new_file)], catch_exceptions=False)
+    assert 0 == result.exit_code, format_result_exception(result)
```
Comment (mohammad-alisafaee) [inline tests/cli/test_datasets.py:1562]: We can use the `project_with_datasets` fixture that already includes 2 datasets and some data files.

**Review by mohammad-alisafaee — CHANGES_REQUESTED** [review]
Thanks Ralf! Please see my comments.

File: renku/infrastructure/gateway/dataset_gateway.py (hunk 1 above)
Comment (mohammad-alisafaee) [reply renku/infrastructure/gateway/dataset_gateway.py:72]: It's also safer to call `.pop(t._p_oid, None)` or raise a `RenkuException`

Comment (Panaetius) [reply renku/infrastructure/gateway/dataset_gateway.py:72]: actually the db already has a `remove_from_cache` method :see_no_evil:

**Review by olevski — APPROVED** [review]

**Review by mohammad-alisafaee — APPROVED** [review]

**Pull request merged**